### PR TITLE
Update to tensorflow.js

### PR DESCRIPTION
**Pull request by cvalenzuela**
Not ready to merge yet!

Following the update to the library (https://github.com/ml5js/ml5-library/pull/117), this PR updates the examples to match the new changes.

Major things are:
- KNN replaced for Transfer Learning: including training and loss elements
- Adding YOLO

I'm using a `/temp_lib` folder to point to the WIP `ml5.js 0.0.4` for now

### CODE REVIEW

File: p5js/ImageClassification/sketch.js [hunk 1]
```diff
@@ -5,12 +5,11 @@
...
 /* ===
 ML5 Example
-ImageNet_Simple
 Simple Image Classification using p5.js
 === */
 
-// Initialize the ImageNet method with the MobileNet model.
-const classifier = new ml5.ImageClassifier('MobileNet');
+// Initialize the Image Classifier method
+const classifier = new ml5.ImageClassifier();
```
Comment (shiffman) [inline p5js/ImageClassification/sketch.js]: Is "MobileNet" the default but the Image classifier could still load other models as well? I think I like the idea of requiring a reference to the pre-trained model to increase transparency. And this also opens up the possibility of creating an "empty" Image Classifier that has to be trained from scratch?

File: p5js/ImageClassification/sketch.js [hunk 2]
```diff
@@ -29,6 +28,6 @@ function imageReady() {
 // When we get the results
 function gotResult(results) {
   // The results are in an array ordered by probability.
-  select('#result').html(results[0].label);
+  select('#result').html(results[0].className);
```
Comment (shiffman) [inline p5js/ImageClassification/sketch.js]: Will "className" be more confusing than "label"? Is it important to use the underlying terminology? (I mean these to be open questions.)

File: p5js/ImageClassification_TransferLearning/sketch.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+// A function to be called when the model has been loaded
+function modelLoaded() {
+  select('#loading').html('Model loaded!');
+}
+
+// Add the current frame from the video to the classifier
+function addImage(label) {
+  classifier.addImage(label);
```
Comment (shiffman) [inline p5js/ImageClassification_TransferLearning/sketch.js]: Note here we are using "label" as a variable name.

File: p5js/ImageClassification_TransferLearning/sketch.js [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+  hiddenUnits: 100,
+  epochs: 20,
+  numClasses: 2,
+  batchSize: 0.4,
+};
+
+function setup() {
+  noCanvas();
+  // Create a video element
+  video = createCapture(VIDEO, onVideoLoaded).parent('videoContainer');
```
Comment (shiffman) [inline p5js/ImageClassification_TransferLearning/sketch.js]: Chaining can sometimes be a little intimidating to beginners.

File: p5js/ImageClassification_TransferLearning/sketch.js [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+=== */
+
+let classifier;
+let video;
+let loss;
+let dogImages = 0;
+let catImages = 0;
+
+// Default options for training
+const options = {
```
Comment (shiffman) [inline p5js/ImageClassification_TransferLearning/sketch.js]: I wonder if we rely on the defaults for our basic example and leave this out. Great to allow this possibility but a little scary looking. I guess now with transfer learning we have to specify the number of classes in advance?

File: p5js/ImageClassification_TransferLearning/sketch.js [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+  buttonA = select('#buttonA');
+  buttonA.mousePressed(function() {
+    addImage('cat');
+    select('#exampleA').html(catImages++);
+  });
+
+  // Add image as dog class
+  buttonB = select('#buttonB');
+  buttonB.mousePressed(function() {
+    addImage('dog');
```
Comment (shiffman) [inline p5js/ImageClassification_TransferLearning/sketch.js]: Yay for string labels!!!

File: p5js/ImageClassification_TransferLearning/sketch.js [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    select('#exampleB').html(dogImages++);
+  });
+
+  // Train Button
+  train = select('#train');
+  train.mousePressed(function() {
```
Comment (shiffman) [inline p5js/ImageClassification_TransferLearning/sketch.js]: These nested callbacks are a little confusing, I can take a look at some point.

File: p5js/YOLO/sketch.js [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+let video;
+let yolo;
+let objects = [];
+
+function setup() {
+  createCanvas(640, 410);
+  video = createCapture(VIDEO, onVideoLoaded);
+  video.hide();
+  select('#start').mousePressed(function() {
```
Comment (shiffman) [inline p5js/YOLO/sketch.js]: Maybe just pass in detect directly? 

```javascript
select('#start').mousePressed(detect);
```

File: p5js/YOLO/sketch.js [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    detect();
+  });
+}
+
+function draw() {
+  image(video, 0, 0, 640, 410);
+  objects.forEach(function(object) {
```
Comment (shiffman) [inline p5js/YOLO/sketch.js]: `forEach` loops aren't commonly used in p5 examples so I might just use a regular for loop or a `for..of` loop?

File: p5js/YOLO/sketch.js [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    text(object.className, object.x*width, object.y*height - 5);
+    noFill();
+    strokeWeight(4);
+    stroke(0,255, 0);
+    rect(object.x*width, object.y*height, object.w*width, object.h*height);
+  });
+}
+
+function onVideoLoaded() {
+  yolo = new ml5.YOLO(video.elt);
```
Comment (shiffman) [inline p5js/YOLO/sketch.js]: A reminder to maybe revisit the question of if we can say things like `yolo = new ml5.YOLO(video)` for integration with p5?  Or is that opening too big of a can of worms? Sorry to be obsessed with this small detail :smile:

https://github.com/ml5js/ml5-library/issues/27

File: p5js/ImageClassification/sketch.js (hunk 1 above)
Comment (cvalenzuela) [reply p5js/ImageClassification/sketch.js]: It is the default because Squeezenet does not currently work in tf.js (See: https://github.com/tensorflow/tfjs-core/tree/master/models)
I also liked specifying the model, we can bring it back!

File: p5js/ImageClassification/sketch.js (hunk 2 above)
Comment (cvalenzuela) [reply p5js/ImageClassification/sketch.js]: I change this because I came across much more references to 'class' and 'class name' rather than 'label' in other examples.

File: p5js/ImageClassification_TransferLearning/sketch.js (hunk 4 above)
Comment (cvalenzuela) [reply p5js/ImageClassification_TransferLearning/sketch.js]: yep, agree! will update then.

File: p5js/ImageClassification_TransferLearning/sketch.js (hunk 5 above)
Comment (cvalenzuela) [reply p5js/ImageClassification_TransferLearning/sketch.js]: These are totally optional, no need to provide them (actually they are the same in the src code). But if you are using more classes and examples in your training you might need to tweak this values a bit.

The defaults expect 2 classes.

File: p5js/YOLO/sketch.js (hunk 8 above)
Comment (cvalenzuela) [reply p5js/YOLO/sketch.js]: ok, I'm still having some issues with the speed (FPS). Although the draw in not waiting for yolo to send results. We can see this tomorrow.

File: p5js/YOLO/sketch.js (hunk 10 above)
Comment (cvalenzuela) [reply p5js/YOLO/sketch.js]: Yes, totally! Just need to update the library.

File: p5js/ImageClassification_TransferLearning/sketch.js (hunk 5 above)
Comment (shiffman) [reply p5js/ImageClassification_TransferLearning/sketch.js]: I wonder if there is some clever way we could wait to configure the number of new classes in the model until right before training.

This is on the borderline of perhaps too much magic behind the scenes, but having a flexible "teachable machine" where the interaction is making up classes and adding training images on the fly might be a useful feature.

Comment (cvalenzuela) [reply p5js/ImageClassification_TransferLearning/sketch.js]: yes, totally. we could have predefined values that are updated based on the number of classes and images. Values we can get running a couple of examples and seeing performance.

**Comment by cvalenzuela** [comment]
Just pushed new updates reflecting the changes to https://github.com/ml5js/ml5-library/pull/117

File: p5js/LSTM_Interactive/index.html [hunk 11]
```diff
@@ -13,7 +13,8 @@
 
   <script src="https://cdnjs.cloudflare.com/ajax/libs/p5.js/0.6.0/p5.min.js"></script>
   <script src="https://cdnjs.cloudflare.com/ajax/libs/p5.js/0.6.0/addons/p5.dom.min.js"></script>
-  <script src="https://unpkg.com/ml5@0.0.3/dist/ml5.js"></script>
+  <!-- <script src="https://unpkg.com/ml5@0.0.3/dist/ml5.js"></script> -->
+  <script src="../../temp_lib/ml5.min.js"></script>
```
Comment (shiffman) [inline p5js/LSTM_Interactive/index.html]: This is b/c we haven't done the release yet?

File: p5js/PoseNet/sketch.js [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+  image(video, 0, 0, w, h);
+
+  // We can call both functions to draw all keypoints and the skeletons
+  drawKeypoints();
+  drawSkeleton();
+}
+
+// A function to draw ellipses over the detected keypoints
+function drawKeypoints() {
+  poses.forEach(pose => {
```
Comment (shiffman) [inline p5js/PoseNet/sketch.js]: I've started using this style/syntax more now myself, but I think for the p5 examples  we should probably stick with more vanilla for loops or `for of` loops.

File: p5js/PoseNet/sketch.js [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+      let y2 = parts[1].position.y;
+      line(x1, y1, x2, y2);
+    });
+  });
+}
+
+// The callback that gets called every time there's an update from the model
+function gotPoses(results) {
+  poses = results;
+  skeletons = poses.map(pose => poseNet.skeleton(pose.keypoints));
```
Comment (shiffman) [inline p5js/PoseNet/sketch.js]: Is this something we can package as a feature of the library? 

```javascript
function gotPoses(poses) {
 skeletons = poseNet.skeleton(poses);
}
```

or

```javascript
function gotPoses(poses, skeletons) {
 // callback passes arrays of both?
 // Does the end user always want the "skeleton" format?
}
```

At the very least we might want to add some comments / explanation of what's happening here.

File: p5js/LSTM_Interactive/index.html (hunk 11 above)
Comment (cvalenzuela) [reply p5js/LSTM_Interactive/index.html]: yep, just temporary

File: p5js/PoseNet/sketch.js (hunk 12 above)
Comment (cvalenzuela) [reply p5js/PoseNet/sketch.js]: Yes, agree. Will update this.

File: p5js/PoseNet/sketch.js (hunk 13 above)
Comment (cvalenzuela) [reply p5js/PoseNet/sketch.js]: totally! I think the later is better, will update then

Comment (shiffman) [reply p5js/PoseNet/sketch.js]: I'm a little confused as to the difference between a `pose` and a `skeleton` but I can take a look at the example and play with it a bit more after we merge!

Comment (cvalenzuela) [reply p5js/PoseNet/sketch.js]: `pose` is an array of keypoints describing the body positions detected. 
```javascript
[ score: 0.9952507019042969, part: "nose", position: { x: 174.63247075574154, y: 226.84211836190062 } , ...]
```

`skeleton` is an array containing arrays of two elements describing connections between two keypoints
```javascript
[[part: "leftShoulder"], [ part: "rightShoulder" ], ... ]
```

Comment (shiffman) [reply p5js/PoseNet/sketch.js]: Ah, that makes sense! Let me know when you are ready for me to merge!

**Comment by cvalenzuela** [comment]
@shiffman, I just updated some new changes. I also added a couple of vanilla js examples.
I think we can merge this and so we can also merge https://github.com/ml5js/ml5-examples/pull/22

**Comment by shiffman** [comment]
Yay!
